gcovr/workers.py

```diff
@@ -0,0 +1,119 @@
+# -*- coding:utf-8 -*-
+
+# This file is part of gcovr <http://gcovr.com/>.
+#
+# Copyright 2013-2018 the gcovr authors
+# This software is distributed under the BSD license.
+
+
+from threading import Thread, Condition
+from contextlib import contextmanager
+
+import sys
+if sys.version_info[0] >= 3:
+    from queue import Queue
+else:
+    from Queue import Queue
+
+
+class LockedDirectories(object):
+    """
+    Class that keeps a list of locked directories
+    """
+    def __init__(self):
+        self.dirs = set()
+        self.cv = Condition()
+
+    def run_in(self, dir_):
+        """
+        Start running in the directory and lock it
+        """
+        self.cv.acquire()
+        while dir_ in self.dirs:
+            self.cv.wait()
+        self.dirs.add(dir_)
+        self.cv.release()
+
+    def done(self, dir_):
+        """
+        Finished with the directory, unlock it
+        """
+        self.cv.acquire()
+        self.dirs.remove(dir_)
+        self.cv.notify_all()
+        self.cv.release()
+
+
+@contextmanager
+def locked_directory(dir_):
+    """
+    Context for doing something in a locked directory
+    """
+    locked_directory.global_object.run_in(dir_)
+    yield
+    locked_directory.global_object.done(dir_)
+
+
+locked_directory.global_object = LockedDirectories()
+
+
+def worker(queue, context, exceptions):
+    """
+    Run work items from the queue until the sentinal
+    None value is hit
+    """
+    while True:
+        work, args, kwargs = queue.get(True)
+        if not work:
+            break
+        kwargs.update(context)
+        try:
+            work(*args, **kwargs)
+        except:  # noqa: E722
+            import sys
+            exceptions.append(sys.exc_info())
+            break
```

Comment:
If one thread exits due to an exception, the others will continue until all work is done (or all have died). So with each exception, processing slows down further. This could e.g. happen if some gcov files have parse errors. I'm also thinking about how this integrates with Ctrl-C/`KeyboardInterrupt` exceptions.

Can we somehow turn this into fail-fast behaviour? E.g. if an exception is caught, the input queue is drained?

Reply:
I agreed. A job for Monday - I'll make sure that if an exception is raised the queue is drained and a sentinel pushed. I think I also need to wrap the call to this in `__main__.py` in a context manager that can terminate threads if an exception is thrown prior to the sentinels being pushed. It would probably be neat to make the pool itself a context.

Reply:
This is done in the latest commit, on an exception the queue is drained and there is code to ensure that the sentinels persist - so that if the drain occurs prior to pool.wait() being called the threads still stop. I've also wrapped the pool in a context manager so that keyboard interrupts are caught and the queue drained from main.

Reply:
Thank you for these updates. At first glance this looks quite good, but a proper review will have to wait until I'm less tired.